docs/core/extensions/httpclient-factory-keyed-di.md

```diff
@@ -0,0 +1,302 @@
+```
+
+### Beware Of Scope Mismatch
+
+While Scoped lifetime is much less problematic for the Named `HttpClient`s (compared to Singleton and Transient pitfalls), it has its own catch.
+
+> [!IMPORTANT]
+> Keyed Scoped lifetime of a specific `HttpClient` instance is bound&mdash;as expected&mdash;to the "ordinary" application scope (for example, incoming request scope) where it was resolved from. However, it does NOT apply to the underlying message handler chain, which is still managed by the `HttpClientFactory`, in the same way it is for the Named clients created directly from factory. `HttpClient`s with the _same_ name, but resolved (within a `HandlerLifetime` timeframe) in two different scopes (for example, two concurrent requests to the same endpoint), can reuse the _same_ `HttpMessageHandler` instance. Which in turn has its own separate scope, as illustrated in the [docs](httpclient-factory.md#message-handler-scopes-in-ihttpclientfactory).
+
+> [!NOTE]
+> The [Scope Mismatch](httpclient-factory-troubleshooting.md#httpclient-doesnt-respect-scoped-lifetime) problem is nasty and long-existing one, and as of .NET 9 still remains [unsolved](https://github.com/dotnet/runtime/issues/47091). From a service injected through the regular DI infra, you would expect all the dependencies to be satisfied from the same scope&mdash;but for the Keyed Scoped `HttpClient`s it is, unfortunately, not the case.
+
+## Keyed Message Handler Chain
+
+For some advanced scenarios, you might want to access `HttpMessageHandler` chain directly, instead of an `HttpClient` object. `HttpClientFactory` provides `IHttpMessageHandlerFactory` interface to create the handlers; and if you enable Keyed DI, then not only `HttpClient`, but also the respective `HttpMessageHandler` chain is registered as a Keyed service:
+
+```csharp
+services.AddHttpClient("foo").AddAsKeyed();
+
+var handler = provider.GetRequiredKeyedService<HttpMessageHandler>("foo");
+var invoker = new HttpMessageInvoker(handler, disposeHandler: false);
+```
+
+## HOW-TO: Switch from Typed approach to Keyed DI
+
+> [!NOTE]
+> We currently recommend using Keyed DI approach instead of Typed clients.
+
+A minimal-change switch from an existing Typed client to a Keyed dependency can look as follows:
+
+```diff
+- services.AddHttpClient<Foo>(        // (1) Typed client
++ services.AddHttpClient(nameof(Foo), // (1) Named client
+      c => { /* ... */ }              // HttpClient configuration
+  //).Configure....
+- );
++ ).AddAsKeyed();                     // (1) + Keyed DI opt-in
+
++ services.AddTransient<Foo>();       // (1) Plain Transient service
+
+  public class Foo(
+-                                      // (2) Implicit ("hidden" Named) dependency
++     [FromKeyedServices(nameof(Foo))] // (2) Explicit Keyed Service dependency
+      HttpClient httpClient) // { ...
+```
+
+In the example:
+
+1. The registration of the Typed client `Foo` is split into:
```

Comment:
We can't use "Foo": https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/f/foo-foobar-fubar?toc=%2Fwriting-style-guide-msft-internal%2Ftoc.json&bc=%2Fwriting-style-guide-msft-internal%2Fbreadcrumb%2Ftoc.json

Reply:
fixed